src/gui/sdlmain.cpp

```diff
@@ -773,17 +786,106 @@ static bool LoadGLShaders(const char *src, GLuint *vertex, GLuint *fragment) {
 }
 #endif
 
-Bitu GFX_SetSize(Bitu width,Bitu height,Bitu flags,double scalex,double scaley,GFX_CallBack_t callback) {
+/* Get the maximum area available for drawing: */
+/* TODO: The current implementation will not let the pixel-perfect mode to  */
+/*       to apply even the simplest 1:2 and 2:1 corrections at the original */
+/*       window resolution, therefore: consider a special case for PP mode. */
+static void GetAvailableArea(int *width, int *height)
+{
+	bool   fixed;
+	double par;
+
+	fixed = false;
+	if (sdl.desktop.fullscreen) {
+		if (sdl.desktop.full.fixed) {
+			*width  = sdl.desktop.full.width;
+			*height = sdl.desktop.full.height;
+			fixed   = true;
+		}
+	}
+	else {
+		if (sdl.desktop.window.width > 0) {
+			*width  = sdl.desktop.window.width;
+			*height = sdl.desktop.window.height;
+			fixed   = true;
+		}
+	}
+	if (!fixed) {
+		par = sdl.draw.par;
+		if (par > 1.0)
+			*height = round( *height * par );
+		if (par < 1.0)
+			*width  = round( *width  / par ); 
```

Comment:
You mean the by-reference parameters available in *C++*? Yes, they are better, but not being a *C++* programmer, I did not even think of them. They are not, however, in all ways better than passing pointers, because pointers let you understand what is going on by looking at the middle of a function, whereas with by-reference parameters you have to look at the declaration. Anyway, I have switched to references.

As to using a *dims* struct, I think it will only make things more clumsy. Inside the function, the programmer will have to write `dims.width` instead `width`, whereas outside he will have to invoke it in a very clumsy way:

1. declare a *dims* local variable,
2. invoke the fucntion,
3. transfer the width and height values from *dims* to the scalar local variables.

Reply:
(not a comment to implementation, just a response to 3 points listed)

Generally, I agree; in some cases it's good to force programmer *writing* code to pass pointers instead of references (specifically when passing pointers to fundamental types), as it leaves a hint to programmer *reading* code, that perhaps this is an output parameter - this is true both in C and C++ (although there are additional things to consider in C++ when desiging function interface, which I won't get into here).

However - in C++11 or newer - point (1) is not valid, e.g.:
```
struct dims
{
	int width;
	int height;
};

constexpr dims double_dimensions(dims size)
{
	size.width  *= 2;
	size.height *= 2;
	return size;
}

// usage:

const auto new_dims = double_dimensions({ 10, 10 });
// C++11 feature used above is called "extended initializer list"; not to be 
// confused with "constructor initialization list" (C++ naming, yeah, I know :/)
int w = new_dims.width;
int h = new_dims.height;

// if dimensions were stored in dims struct from the beginning:
sdl.foobar.window_size = double_dimensions(sdl.foobar.window_size);
```
Structs can be initialized in-place, without an additional local variable. Functions written this way have a bonus of being completely-pure, so can be declared constexpr.

Minuses:
- values still need to be transferred from temporary struct to the variables (but it could be avoided if the correlated values were stored in a struct as well); IIRC it's possible to avoid this, but code readability would suffer

Pluses:
- code is easier to understand (it's clear what's input and what's output)
- it's easier to write unit tests
- pure function can be declared `constexpr`, which allows compiler to evaluate it at compile time (when possible), so it might be faster as well